sound/soc/sof/sof-audio.c

```diff
@@ -321,14 +322,15 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
 	swidget->prepared = true;
 
 sink_prepare:
+	memcpy(&params, pipeline_params, sizeof(*pipeline_params));
 	/* prepare all widgets in the sink paths */
 	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
 		if (!widget_in_list(list, p->sink))
 			continue;
 		if (!p->walking && p->sink->dobj.private) {
 			p->walking = true;
 			ret = sof_prepare_widgets_in_path(sdev, p->sink,  fe_params,
-							  platform_params, pipeline_params, dir,
+							  platform_params, &params, dir,
```

Comment:
I don't see how this achieves the stated goal.
You copy the `pipeline_params` to a local `params` struct and pass that pointer recursively to the next pipeline, which would make sense _if_ the `pipeline_params` is modified at all down the call chain, but I just can not see that.

For me this is a NOP change with an extra local variable and a memcpy. Or do I miss something?

Reply:
Right. We should move the memcpy into the loop. Thanks @ranj063 @ujfalusi

Reply:
@libinyang even though the change looks correct now, I'd be cautious merging this without a SOF PR that shows that it actually works as expected

Reply:
How did you find the issue and how did you test your patch @libinyang ?

Reply:
@plbossart When I developed the WoV feature, I need to change the hw_params on the runtime. So, I reviewed the hw_params related code in the sof and found that changing the hw_params will impact other unrelated pipeline widgets. I haven't any topology to test this patch. I just created this patch based on the logic. If we need to test this PR, I can create some toplogies to do some tests.

Reply:
I would recommend we first focus on what we need to do as part of our objectives.
We have many other things that are broken and need urgent attention.

We can revisit this when we have more time and a clearer view of what happens when hw_params are changed. I would bet it's already broken in a single linear pipeline.

Reply:
maybe keep it as draft and we'll think later about this one?